kvserver: broken tests in Bazel

We are making a lot of progress in the migration from make to Bazel. Before we can complete our migration, we need to make sure that all tests in-tree continue to pass when run inside of the Bazel sandbox. The following tests are broken when run in the Bazel sandbox:

    pkg/base:base_test
    pkg/internal/codeowners:codeowners_test
    pkg/internal/team:team_test
    pkg/kv/kvserver:kvserver_test
    pkg/server:server_test
    pkg/server/debug/goroutineui:goroutineui_test
    

Please help us by doing the following:

* Reproduce the failure locally; for example, with bazel test `bazel test pkg/base:base_test`, and consult the test logs to see what went wrong
* Identify the root cause of the failure
* If you are able, repair the failure such that the test passes, and submit a PR removing the `broken_in_bazel` tag from the test (see the linked documentation below)
* If not, take your findings to #bazel in Slack, and we’ll work with you to figure out how we can get the tests working

Helpful documentation:
* [How to ensure your tests can run in the Bazel sandbox (consider starting here!)](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1553172016/How+to+ensure+your+tests+can+run+in+the+Bazel+sandbox)
* [How to ensure your code builds with Bazel](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1380090083/How+to+ensure+your+code+builds+with+Bazel)
* [A user's guide to Bazel](https://docs.bazel.build/versions/4.0.0/guide.html)

Works in Docker, and I get the actual test failure now. 👍 Will fix.

```
--- FAIL: TestValidateAddrs (5.14s)
    --- FAIL: TestValidateAddrs/5/{:postgresql__:http__:postgresql_} (0.00s)
        addr_validation_test.go:169: expected error "", got invalid --advertise-addr: address tcp/postgresql: unknown port
```

--- a/.github/BUILD.bazel
+++ b/.github/BUILD.bazel
@@ -1,0 +1,3 @@
+exports_files([
+    "CODEOWNERS",
+])

--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -188,3 +188,7 @@ gazelle(
     name = "gazelle",
     prefix = "github.com/cockroachdb/cockroach",
 )
+
+exports_files([
+    "TEAMS.yaml",
+])

--- a/build/bazelbuilder/Dockerfile
+++ b/build/bazelbuilder/Dockerfile
@@ -15,6 +15,7 @@ RUN apt-get update \
     libncurses-dev \
     libtinfo-dev \
     make \
+    netbase \
     python3 \
     unzip \
  && update-alternatives --install /usr/bin/clang clang /usr/bin/clang-10 100 \

--- a/build/teamcity-bazel-support.sh
+++ b/build/teamcity-bazel-support.sh
@@ -1,1 +1,1 @@
-BAZEL_IMAGE=cockroachdb/bazel:20210311-171107
+BAZEL_IMAGE=cockroachdb/bazel:20210315-115417

--- a/pkg/BUILD.bazel
+++ b/pkg/BUILD.bazel
@@ -351,8 +351,8 @@ ALL_TESTS = [
 ]
 
 # These suites run only the tests with the appropriate "size" (excepting those
-# tagged "broken_in_bazel") [1]. Note that tests have a default timeout
-# depending on the size [2].
+# tagged "broken_in_bazel" or "flaky") [1]. Note that tests have a default
+# timeout depending on the size [2].
 
 # [1] https://docs.bazel.build/versions/master/be/general.html#test_suite
 # [2] https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes-tests
@@ -361,6 +361,7 @@ test_suite(
     name = "small_tests",
     tags = [
         "-broken_in_bazel",
+        "-flaky",
         "small",
     ],
     tests = ALL_TESTS,
@@ -370,6 +371,7 @@ test_suite(
     name = "medium_tests",
     tags = [
         "-broken_in_bazel",
+        "-flaky",
         "medium",
     ],
     tests = ALL_TESTS,
@@ -379,6 +381,7 @@ test_suite(
     name = "large_tests",
     tags = [
         "-broken_in_bazel",
+        "-flaky",
         "large",
     ],
     tests = ALL_TESTS,
@@ -388,6 +391,7 @@ test_suite(
     name = "enormous_tests",
     tags = [
         "-broken_in_bazel",
+        "-flaky",
         "enormous",
     ],
     tests = ALL_TESTS,

--- a/pkg/base/BUILD.bazel
+++ b/pkg/base/BUILD.bazel
@@ -53,7 +53,6 @@ go_test(
         "node_id_test.go",
         "store_spec_test.go",
     ],
-    tags = ["broken_in_bazel"],
     deps = [
         ":base",
         "//pkg/roachpb",

--- a/pkg/cmd/generate-test-suites/main.go
+++ b/pkg/cmd/generate-test-suites/main.go
@@ -42,8 +42,8 @@ ALL_TESTS = [`)
 	fmt.Println(`]
 
 # These suites run only the tests with the appropriate "size" (excepting those
-# tagged "broken_in_bazel") [1]. Note that tests have a default timeout
-# depending on the size [2].
+# tagged "broken_in_bazel" or "flaky") [1]. Note that tests have a default
+# timeout depending on the size [2].
 
 # [1] https://docs.bazel.build/versions/master/be/general.html#test_suite
 # [2] https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes-tests`)
@@ -54,6 +54,7 @@ test_suite(
     name = "%[1]s_tests",
     tags = [
         "-broken_in_bazel",
+        "-flaky",
         "%[1]s",
     ],
     tests = ALL_TESTS,

--- a/pkg/internal/codeowners/BUILD.bazel
+++ b/pkg/internal/codeowners/BUILD.bazel
@@ -17,9 +17,13 @@ go_test(
     name = "codeowners_test",
     size = "small",
     srcs = ["codeowners_test.go"],
+    data = [
+        "//:TEAMS.yaml",
+        "//.github:CODEOWNERS",
+    ],
     embed = [":codeowners"],
-    tags = ["broken_in_bazel"],
     deps = [
+        "//pkg/build/bazel",
         "//pkg/internal/team",
         "@com_github_stretchr_testify//require",
     ],

--- a/pkg/internal/codeowners/codeowners_test.go
+++ b/pkg/internal/codeowners/codeowners_test.go
@@ -11,12 +11,30 @@
 package codeowners
 
 import (
+	"log"
 	"strings"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/build/bazel"
 	"github.com/cockroachdb/cockroach/pkg/internal/team"
 	"github.com/stretchr/testify/require"
 )
+
+func init() {
+	if bazel.BuiltWithBazel() {
+		codeOwnersFile, err := bazel.Runfile(".github/CODEOWNERS")
+		if err != nil {
+			log.Fatal(err)
+		}
+		DefaultCodeOwnersLocation = codeOwnersFile
+
+		teamFile, err := bazel.Runfile("TEAMS.yaml")
+		if err != nil {
+			log.Fatal(err)
+		}
+		team.DefaultTeamsYAMLLocation = teamFile
+	}
+}
 
 func TestMatch(t *testing.T) {
 	owners := `

--- a/pkg/internal/team/BUILD.bazel
+++ b/pkg/internal/team/BUILD.bazel
@@ -15,7 +15,12 @@ go_test(
     name = "team_test",
     size = "small",
     srcs = ["team_test.go"],
+    data = [
+        "//:TEAMS.yaml",
+    ],
     embed = [":team"],
-    tags = ["broken_in_bazel"],
-    deps = ["@com_github_stretchr_testify//require"],
+    deps = [
+        "//pkg/build/bazel",
+        "@com_github_stretchr_testify//require",
+    ],
 )

--- a/pkg/internal/team/team_test.go
+++ b/pkg/internal/team/team_test.go
@@ -12,10 +12,22 @@ package team
 
 import (
 	"bytes"
+	"log"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/build/bazel"
 	"github.com/stretchr/testify/require"
 )
+
+func init() {
+	if bazel.BuiltWithBazel() {
+		teamFile, err := bazel.Runfile("TEAMS.yaml")
+		if err != nil {
+			log.Fatal(err)
+		}
+		DefaultTeamsYAMLLocation = teamFile
+	}
+}
 
 func TestLoadTeams(t *testing.T) {
 	yamlFile := []byte(`

--- a/pkg/kv/kvserver/BUILD.bazel
+++ b/pkg/kv/kvserver/BUILD.bazel
@@ -285,7 +285,7 @@ go_test(
     data = glob(["testdata/**"]),
     embed = [":kvserver"],
     shard_count = 16,
-    tags = ["broken_in_bazel"],
+    tags = ["flaky"],
     deps = [
         "//pkg/base",
         "//pkg/cli/exit",

--- a/pkg/server/BUILD.bazel
+++ b/pkg/server/BUILD.bazel
@@ -275,7 +275,6 @@ go_test(
     data = glob(["testdata/**"]),
     embed = [":server"],
     shard_count = 16,
-    tags = ["broken_in_bazel"],
     deps = [
         "//pkg/base",
         "//pkg/build",

--- a/pkg/server/debug/goroutineui/BUILD.bazel
+++ b/pkg/server/debug/goroutineui/BUILD.bazel
@@ -15,9 +15,10 @@ go_test(
     name = "goroutineui_test",
     size = "small",
     srcs = ["dump_test.go"],
+    data = ["@go_sdk//:files"],
     embed = [":goroutineui"],
-    tags = ["broken_in_bazel"],
     deps = [
+        "//pkg/build/bazel",
         "//pkg/util/leaktest",
         "@com_github_stretchr_testify//assert",
     ],

--- a/pkg/server/debug/goroutineui/dump_test.go
+++ b/pkg/server/debug/goroutineui/dump_test.go
@@ -15,9 +15,16 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/build/bazel"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/stretchr/testify/assert"
 )
+
+func init() {
+	if bazel.BuiltWithBazel() {
+		bazel.SetGoEnv()
+	}
+}
 
 func TestDumpHTML(t *testing.T) {
 	defer leaktest.AfterTest(t)()

